### ISSUE TITLE
Incorrect 'unsubscriptable-object' reported

### ISSUE THREAD

**Issue by qsmy41** [PyCQA/pylint#3637]
### Steps to reproduce
Running Pylint on the following code will produce the error on the last line.
```python
import statsmodels.api as sm

data = sm.datasets.longley.load_pandas().data

properties = ['GNP', 'POP']
specificData = data[properties]

```

### Current behavior
Running Pylint gives 
```
example.py|6 col 16 error| E1136: Value 'data' is unsubscriptable (unsubscriptable-object)
example.py|6 col 16 error| unsubscriptable-object: Value 'data' is unsubscriptable
```

### Expected behavior
No errors/warnings should be produced.

### pylint --version output
pylint 2.5.2
astroid 2.4.1
Python 3.7.7 (default, Apr 14 2020, 11:51:19)
[Clang 11.0.0 (clang-1100.0.33.12)]

**Comment by beauxq** [comment]
Same problem with this:
```
import numpy as np

ARRAY = np.array([[1, 2, 3]]).T

print(ARRAY[0])
```
`Value 'ARRAY' is unsubscriptable (unsubscriptable-object)`
There should be no error.

**Comment by acxz** [comment]
can confirm also getting this issue with

pylint 2.6.0
astroid 2.4.2
Python 3.8.6

related issues: https://github.com/PyCQA/pylint/issues/3139
https://github.com/PyCQA/pylint/issues/2416

**Comment by Cielquan** [comment]
Have a similar problem with type hints.

This LoC 
```python
EnvVarTypes = Union[str, int, float, bool, None]
```
gives `E1136: Value 'Union' is unsubscriptable (unsubscriptable-object)`.

This LoC 
```python
def validate(self, line: str, _: GitCommit) -> List[Optional[RuleViolation]]:
```
gives `E1136: Value 'Optional' is unsubscriptable (unsubscriptable-object)`.

BUT those errors only occur in with
```console
$ pylint --version
pylint 2.6.0
astroid 2.4.2
Python 3.9.0 (default, Oct  6 2020, 02:47:17) 
[GCC 7.5.0]
```
and not with
```console
$ pylint --version
pylint 2.6.0
astroid 2.4.2
Python 3.8.6 (default, Oct  6 2020, 03:22:36) 
[GCC 7.5.0]
```

**Comment by leycec** [comment]
This also breaks [PEP 560](https://www.python.org/dev/peps/pep-0560) and [PEP 585](https://www.python.org/dev/peps/pep-0585). See [this comment at the related issue #3882](https://github.com/PyCQA/pylint/issues/3882#issuecomment-731515123).

**Comment by hippo91** [comment]
@beauxq thanks for your comment. It seems your issue is not linked to the original one. Morevoer it should have been fixed through PyCQA/astroid#831.

**Comment by hippo91** [comment]
@Cielquan thanks for your report. However your issue is not linked to the original one but it is due to the fact `pylint` is not yet supporting `python3.9`. We are working on it.

**Comment by hippo91** [comment]
To any people thinking adding a comment here, please do only if your problem seems to be linked to `panda`'s support in `pylint`. Otherwise please open another issue. Thanks.

**Comment by MiltiadisKoutsokeras** [comment]
Another case that breaks our code checking CI/CD:

```python
"""Pylint unsubscriptable-object example"""
import pandas as pd

result = pd.DataFrame()
cols = ['a', 'b', 'c', 'd']
CSVDATA = 'data.csv'
df = pd.read_csv(CSVDATA)
df['d'] = [CSVDATA.split('.', maxsplit=1)[0]] * df.shape[0]
result = pd.concat([result, df[cols]], axis=0)
print(result)
```

Pylint reports:

```text
************* Module pylint_wrong
pylint_wrong.py:9:28: E1136: Value 'df' is unsubscriptable (unsubscriptable-object)

------------------------------------------------------------------
Your code has been rated at 3.75/10 (previous run: 3.75/10, +0.00)
```

**Comment by wclark** [comment]
I ran into this issue myself, and after looking over some examples here and in related/duplicated tickets, I have a hunch that the cause is that methods like `drop` (or in my case, `set_index`) don't always return a DataFrame -- sometimes they return None, per the Pandas API.  If you force the methods to return None (by passing an `inplace=True` parameter) and restructure your code accordingly, this error seems to go away.

I'd argue this is still something to fix in Pylint, because these Pandas methods ONLY seem to return None when `inplace=True` is passed -- which Pylint could (and arguably should) easily check.

**Comment by nathanjmcdougall** [comment]
Note that now the original example (and similar cases) now give `invalid-name` in addition to E1136:

```Python
import statsmodels.api as sm

data = sm.datasets.longley.load_pandas().data

properties = ['GNP', 'POP']
specificData = data[properties]
```

```Shell
************* Module yetanotherbug
yetanotherbug.py:4:0: C0103: Constant name "data" doesn't conform to UPPER_CASE naming style (invalid-name)
yetanotherbug.py:7:15: E1136: Value 'data' is unsubscriptable (unsubscriptable-object)
```
Also, [juliangilbey](https://github.com/juliangilbey)'s example is now fixed with the latest versions of pylint and pandas.

**Comment by scorphus** [comment]
I cannot reproduce this issue with the above snippets.

However, the following snippet reproduces the issue:

```python
"""Pylint unsubscriptable-object example"""

import random

import pandas as pd


data = []
for n in range(random.randint(0, 10)):
    data.append({"number": n**2})

df = pd.DataFrame(data) if len(data) > 0 else None

if df is None:
    raise ValueError("DataFrame is empty")

print(df["number"].sum())

df = pd.DataFrame(df)  # This line clears the error (and is my workaround for now)
print(df["number"].sum())
```

```
pylint 3.2.6
astroid 3.2.4
Python 3.11.9 (main, May 23 2024, 10:16:49) [Clang 15.0.0 (clang-1500.3.9.4)]
```

Thanks for taking the time to look into this.

**Comment by PhanAlexandre** [comment]
I have also the same errors but not with all DataFrame. Only some dataframe cause problem with drop or other function
I thinks this is related to all mentioned bug.
```
import pandas as pd

data = [["a",1,"","foo"],
        ["b",2,"","bar"],
        ["c",3,"","test"],
        ["a",4,"","bar"],
        ["a",5,"","foo"],
        ["a",5,"v","foo"]
        ]

data2 = {"a":[1,2,3],
         "b":["a","b","c"],
         "c":["foo","bar","foobar"]}

df = pd.DataFrame(data, columns=["a","b","c","d"])  #False positive error  E1136/1137 after .drop
df = df.drop(columns="a")
df["b"] = df['c']

df1 = pd.read_excel("some_path.xlsx")  #no error, same for some other like read_parquet/csv...
df1 = df1.drop(columns="a")  
df1["b"] = df1['c']

df2 = pd.DataFrame.from_records(data, columns=["a","b","c","d"]) #No error
df2 = df2.drop(columns="a")
df2["b"] = df2['c']

df3 = pd.DataFrame.from_dict(data2) #False positive error  E1136/1137 after the .drop
df3 = df3.drop(columns="a")
df3["b"] = df3['c']
```
```
************* Module import pandas as pd
import pandas as pd.py:17:0: E1137: 'df' does not support item assignment (unsupported-assignment-operation)
import pandas as pd.py:17:10: E1136: Value 'df' is unsubscriptable (unsubscriptable-object)
import pandas as pd.py:30:0: E1137: 'df3' does not support item assignment (unsupported-assignment-operation)
import pandas as pd.py:30:11: E1136: Value 'df3' is unsubscriptable (unsubscriptable-object)

-----------------------------------
Your code has been rated at 0.00/10
```

```
pylint 3.3.5
astroid 3.3.8
Python 3.12.5 | packaged by Anaconda, Inc. | (main, Sep 12 2024, 18:18:29) [MSC v.1929 64 bit (AMD64)]
```
